Add QDM workflow components

brews opened the pull request:
This PR:

- Adds `dodola find-qdm-ryw az://simulation.zarr --out qdm-year-range.json`, using new `dodola.services.find_qdm_rollingyearwindow()`. This gets the first and last year for a simulation to adjust with quantile delta mapping (QDM) bias correction. The output years account for a ± 10 year window and 15 days at the start/end of the edge years. The command outputs to a JSON file to be read by Argo Workflows for its parallel "fan-out" step in the bias correction workflow. The JSON file looks like `{"firstyear": <int earliest year>, "lastyear": <int latest year>}`.

Close #66

dgergel commented:
@brews can you resolve the conflicts? Then I'll review. Thanks!

brews commented:
> @brews can you resolve the conflicts? Then I'll review. Thanks!

Yeah, thanks. On it.

brews commented:
@dgergel Merge conflicts are resolved!

dgergel commented on dodola/tests/test_services.py:
```diff
@@ -84,6 +91,114 @@ def domain_file(request):
+    """Test that train_qdm outputs store giving sdba.adjustment.QuantileDeltaMapping
+
+    Checks that output is consistent if we do "temperature" or "precipitation"
+    QDM kinds.
```
update this to be additive or multiplicative kinds

brews replied:
Cool. I got this in `fbe8d03f`. Marking resolved.

dgergel reviewed (changes requested):
@brews thanks so much for putting this together! Overall I think it looks great. My main comment is that all of the `kind` terminology should be `additive` or `multiplicative` rather than variable-dependent since we will likely be treating DTR as a ratio (e.g. multiplicative) variable. 

My other concern is that the rolling year window function is a bit messy, so I was confused as to whether it was actually working correctly - have you tested that it has the correct number of years that we want? Also, it was hard to tell if it's handling edge cases at the very end of the century, e.g. 2090-on - maybe make the documentation and testing on this a bit more clear?

brews commented on dodola/cli.py:
```diff
@@ -36,6 +36,61 @@ def dodola_cli(debug):
+    help="URL to write NetCDF4 with adjusted simulation year to",
+)
+def apply_qdm(simulation, qdm, year, variable, out):
+    """Adjust simulation year with QDM, outputting to local NetCDF4 file"""
```
Fixed in `2d509943`. Resolving.

brews commented on dodola/cli.py:
```diff
@@ -36,6 +36,61 @@ def dodola_cli(debug):
+@click.option(
+    "--simulation", "-s", required=True, help="URL to simulation store to adjust"
+)
+@click.option("--qdm", "-q", required=True, help="URL to trained QDM store")
```
Fixed in `0ad4f527`. Resolving.

brews commented on dodola/cli.py:
```diff
@@ -36,6 +36,61 @@ def dodola_cli(debug):
+    "--kind",
+    "-k",
+    required=True,
+    type=click.Choice(["temperature", "precipitation"], case_sensitive=True),
```
Excellent suggestion. I hadn't thought of that. I fixed this more broadly in `fbe8d03f`.

Marking this as resolved.

brews commented on dodola/cli.py:
```diff
@@ -36,6 +36,61 @@ def dodola_cli(debug):
+)
+@click.option("--out", "-o", required=True, help="URL to write QDM model to")
+def train_qdm(historical, reference, out, variable, kind):
+    """Train Quantile Delta Mapping (QDM) and output to storage"""
```
Fixed in `156447a5`. Marking this as resolved.

brews commented on dodola/services.py:
```diff
@@ -28,6 +33,100 @@ def service_logger(*args, **kwargs):
+        ValueError(f"kind must be {set(kind_map.keys())}, got {kind}")
+
+    qdm = train_quantiledeltamapping(
+        historical=hist, reference=ref, variable=variable, kind=kind_map[kind]
```
Ah, I see. You're saying it's different from the order usually give in `xclim`. In `e5f806dc` I changed the signature in the core and services to reflect this.

Marking as resolved.

brews commented on dodola/services.py:
```diff
@@ -28,6 +33,100 @@ def service_logger(*args, **kwargs):
+        fsspec-compatible path or URL pointing to NetCDF4 file where the
+        QDM-adjusted simulation data will be written.
+    """
+    sim_df = storage.read(simulation)
```
Okay. Fixed in `10c8ec49`. Marking as resolved.

brews commented on dodola/tests/test_core.py:
```diff
@@ -0,0 +1,147 @@
+    ],
+)
+def test_qdm_rollingyearwindow(in_dts, goalyears):
+    """Test qdm_rollingyearwindow ensures account for have 15 day buffer in edge years"""
```
Thank you! I think I cleaned this up in `9b7e8023`. Marking this as resolved.

brews commented on dodola/tests/test_core.py:
```diff
@@ -0,0 +1,147 @@
+        pytest.param(
+            ("2015-01-01", "2100-02-01"), (2026, 2089), id="early start, late end"
+        ),
+    ],
```
I added a new test case in `7abac70a` that tests if we input a simulation ds with "time" from "2004-12-15" through "2111-01-15", the earliest year we would apply QDM for is 2015 and the latest year is 2100.

I think this addresses your concern, @dgergel ? Happy to discuss this further.

dgergel replied:
We are still bias correcting the historical period too though - we are actually going to be bias correcting the historical model data starting at 1950. I think for now we can start at 1980 and then add in the 1950-1980 years later.

dgergel commented:
@brews updating my earlier comment: I think that the `find_qdm_rollingyearwindow()` still needs some work. I took a much closer look at it and I believe the problem is that it's still only going to fan out a smaller number of years than we need, unless I'm misunderstanding something. For example, if we pass in a ds with time 1980-2100, it will fan out 1991 - 2089, which in fact we want 1980-2100 fanned out. And to bias correct those years, we need data from 12-17-1979 to the end of 2100.

dgergel commented:
@brews the goal is to have bias corrected and downscaled output from 1980-2100 like you said. For the edge years, we'd just have fewer years and/or days. Since we have historical model data before 1980, this won't be an issue for the earlier end. For 2089-2100 years, we'll have fewer years in our rolling window. And same for the number of days.

brews commented:
Okay, @dgergel, I think I addressed all of your issues, including removing everything related to `find_qdm_rollingyearwindow()`, et al..

I think this is ready for a re-review!

dgergel reviewed (approved):
@brews LGTM. Thanks for all of your work on this!
